### PR TITLE
Move more methods to visitable interface

### PR DESCRIPTION
**Pull request by mugling**
Moves more functions to `visitable` and then provides an optimised version of `has_amount` that avoids checking every possible item.

Currently charges or items can be stolen from within guns and magazines (for example crafting will remove ammo or a gunmod from a gun). 9dd696b provides a partial fix to ensure that the visitor pattern excludes these. A full fix will therefore occur once `use_charges` and related functions are also migrated.

A further PR will provide specialisations of `charges_of` and `amount_of` supporting `PSEUDO` items so that `vehicle_selector` will correctly detect hotplates etc. Once this is complete idioms such as  `g->u.nearby(4).has_amount("welder", 1)` become possible. Such an invocation would check for either a welder in the player inventory or a welding rig within 4 tiles stopping traversal as soon as it was found.

### CODE REVIEW

File: src/visitable.h [hunk 1]
```diff
@@ -30,14 +30,13 @@ class visitable
...
-        VisitResponse visit_items_with_parent_const(
-            const std::function<VisitResponse( const item *, const item * )> &func ) const;
+        VisitResponse visit_items( const std::function<VisitResponse( item *, item * )> &func );
+        VisitResponse visit_items( const std::function<VisitResponse( const item *, const item * )> &func )
+        const;
 
         /** Lightweight version which provides only the current node */
         VisitResponse visit_items( const std::function<VisitResponse( item * )> &func );
-        VisitResponse visit_items_const( const std::function<VisitResponse( const item * )> &func ) const;
+        VisitResponse visit_items( const std::function<VisitResponse( const item * )> &func ) const;
```
Comment (Coolthulhu) [inline src/visitable.h]: By the way, this one could have a `const item &` overload, as I don't think it is invoked on `nullptr`s.

File: src/visitable.h [hunk 2]
```diff
@@ -60,6 +59,29 @@ class visitable
...
+         * Count items matching id including both this instance and any contained items
+         * @param pseudo whether pseudo-items (from map/vehicle tiles, bionics etc) are considered
+         * @param limit stop searching after this many matches
+         * @note items must be empty to be considered a match
+         */
+        int amount_of( const std::string &what, bool pseudo = true, int limit = INT_MAX ) const;
+
+        /** Check instance provides at least qty of an item (@see amount_of) */
+        bool has_amount( const std::string &what, int qty, bool pseudo = true ) const {
+            return amount_of( what, pseudo, qty ) == qty;
```
Comment (Coolthulhu) [inline src/visitable.h]: Not wrong, but I recall some old variant of `amount_of` understanding early quitting. Crafting inventories can get huge, so early quitting makes sense.

File: src/visitable.h (hunk 1 above)
Comment (mugling) [inline src/visitable.h]: > as I don't think it is invoked on nullptrs.

Yes, the implementation guarantees that.

> By the way, this one could have a const item & overload

I'm still considering the idea of full templates so I'll defer that decision for now if thats ok

File: src/visitable.cpp [hunk 3]
```diff
@@ -471,6 +517,104 @@ std::list<item> visitable<vehicle_selector>::remove_items_with( const
...
 
+template <typename T>
+static long charges_of_internal( const T& self, const itype_id& id )
+{
+    long qty = 0;
+
+    self.visit_items( [&]( const item *e ) {
+        if( e->is_tool() ) {
+            // for tools we also need to check if this item is a subtype of the required id
+            if( e->typeId() == id || ( e->is_tool() && e->type->tool->subtype == id ) ) {
```
Comment (BevapDin) [inline src/visitable.cpp]: `e->is_tool` has already been checked 2 lines above.

File: src/player.cpp [hunk 4]
```diff
@@ -9190,31 +9190,6 @@ bool player::is_waterproof( const std::bitset<num_bp> &parts ) const
...
-    return (amount_of(it) >= quantity);
-}
-
-int player::amount_of(const itype_id &it) const
-{
-    if (it == "toolset" && has_active_bionic("bio_tools")) {
-        return 1;
-    }
-    if (it == "apparatus") {
-        return ( has_items_with_quality("SMOKE_PIPE", 1, 1) ? 1 : 0 );
```
Comment (illi-kun) [inline src/player.cpp]: @mugling removing this broke the smoking pipes. Could you give me a clue how to fix this simply? (or feel free to fix it, if you have time and desire to do so).

**Pull request merged**